File: glcm_cpu.py
"""
Pipeline CPU sequencial para extração de features GLCM + clustering.
Baseline para comparação com implementações CUDA.
"""

import time
import numpy as np
from skimage.feature import graycomatrix, graycoprops
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
from sklearn.preprocessing import StandardScaler

N_LEVELS = 64        # níveis de quantização (6 bits)
PATCH_SIZE = 64      # janela de extração
STEP = 32            # passo (overlap 50 %)
DISTANCES = [1]
ANGLES = [0, np.pi / 4, np.pi / 2, 3 * np.pi / 4]
N_CLUSTERS = 3
HARALICK_PROPS = ["contrast", "dissimilarity", "homogeneity", "energy", "correlation"]


def quantize(image: np.ndarray) -> np.ndarray:
    """uint16 [0,65535] → uint8 [0,63] via deslocamento de 10 bits."""
    return (image >> 10).astype(np.uint8)


def extract_patches(image_q: np.ndarray):
    H, W = image_q.shape
    patches = []
    for y in range(0, H - PATCH_SIZE + 1, STEP):
        for x in range(0, W - PATCH_SIZE + 1, STEP):
            patches.append(image_q[y : y + PATCH_SIZE, x : x + PATCH_SIZE])
    return patches


def compute_features_cpu(patches: list) -> np.ndarray:
    """Retorna [n_patches, 20] — 5 props × 4 ângulos por patch."""
    features = []
    for patch in patches:
        glcm = graycomatrix(
            patch,
            distances=DISTANCES,
            angles=ANGLES,
            levels=N_LEVELS,
            symmetric=True,
            normed=True,
        )
        feat = np.concatenate(
            [graycoprops(glcm, prop).ravel() for prop in HARALICK_PROPS]
        )
        features.append(feat)
    return np.array(features, dtype=np.float64)


def run_cpu_pipeline(image_path: str) -> dict:
    image = np.load(image_path)

    t0 = time.perf_counter()

    image_q = quantize(image)
    patches = extract_patches(image_q)

    t1 = time.perf_counter()

    features = compute_features_cpu(patches)

    t2 = time.perf_counter()

    scaler = StandardScaler()
    features_s = scaler.fit_transform(features)
    km = KMeans(n_clusters=N_CLUSTERS, random_state=42, n_init=5)
    labels = km.fit_predict(features_s)
    sil = float(silhouette_score(features_s, labels))

    t3 = time.perf_counter()

    return {
        "total_time_s": t3 - t0,
        "glcm_time_s": t2 - t1,
        "clustering_time_s": t3 - t2,
        "silhouette_score": sil,
        "n_patches": len(patches),
    }


if __name__ == "__main__":
    import sys

    path = sys.argv[1] if len(sys.argv) > 1 else "./data/image_512x512.npy"
    r = run_cpu_pipeline(path)
    print(f"CPU pipeline — {path}")
    for k, v in r.items():
        print(f"  {k}: {v:.4f}" if isinstance(v, float) else f"  {k}: {v}")

File: glcm_cpu.c
/*
 * glcm_cpu.c  —  Pipeline CPU sequencial: baseline para benchmark.
 *
 * Substitui glcm_cpu.py. Implementa em C puro (sem Python/numpy/skimage):
 *   1. Leitura da imagem binária
 *   2. Quantização uint16 → uint8 (6 bits)
 *   3. Extração de patches
 *   4. Cálculo de GLCM (4 ângulos, simétrica, normalizada)
 *   5. Extração de 5 propriedades de Haralick por ângulo (20 features/patch)
 *   6. Normalização Z-score
 *   7. KMeans (K=3) + Silhouette Score
 *
 * Compile:
 *   gcc -O3 -o glcm_cpu glcm_cpu.c -lm
 *
 * Uso:
 *   ./glcm_cpu ./data/image_512x512.bin
 */

#include "glcm_common.h"

/* ── Ângulos: deslocamentos (dx, dy) para 0°, 45°, 90°, 135° ────────────── */
static const int DX[N_ANGLES] = { 1,  1, 0, -1 };
static const int DY[N_ANGLES] = { 0,  1, 1,  1 };

/* ── Cálculo de GLCM para um patch e um ângulo ──────────────────────────── *
 * glcm_out: float[N_LEVELS * N_LEVELS] zerado antes de chamar             */
static void compute_glcm_patch(const uint8_t *patch, int pH, int pW,
                                int angle_idx, float *glcm_out) {
    int dx = DX[angle_idx], dy = DY[angle_idx];
    long long total = 0;

    for (int y = 0; y < pH; y++) {
        for (int x = 0; x < pW; x++) {
            int nx = x + dx, ny = y + dy;
            if (nx < 0 || nx >= pW || ny < 0 || ny >= pH) continue;
            int v1 = patch[y * pW + x];
            int v2 = patch[ny * pW + nx];
            glcm_out[v1 * N_LEVELS + v2] += 1.0f;
            glcm_out[v2 * N_LEVELS + v1] += 1.0f;   /* simétrica */
            total += 2;
        }
    }

    /* Normalizar */
    if (total > 0) {
        float inv = 1.0f / (float)total;
        for (int i = 0; i < GLCM_ELEMS; i++)
            glcm_out[i] *= inv;
    }
}

/* ── Pipeline CPU completo ──────────────────────────────────────────────── */
RunResult run_cpu_pipeline(const char *image_path) {
    RunResult res = {0};

    int H, W;
    uint16_t *img16 = load_image_bin(image_path, &H, &W);
    if (!img16) { fprintf(stderr, "Falha ao carregar %s\n", image_path); return res; }

    double t0 = now_sec();

    /* ── Quantização ─────────────────────────────────────────────────────── */
    int N = H * W;
    uint8_t *img8 = (uint8_t *)malloc(N * sizeof(uint8_t));
    quantize(img16, img8, N);
    free(img16);

    /* ── Extração de patches ─────────────────────────────────────────────── */
    int n_patches = count_patches(H, W);
    uint8_t *patches = (uint8_t *)malloc((size_t)n_patches * PATCH_SIZE * PATCH_SIZE);
    int actual_patches = 0;
    extract_patches(img8, H, W, patches, &actual_patches);
    free(img8);
    n_patches = actual_patches;
    res.n_patches = n_patches;

    double t1 = now_sec();

    /* ── GLCM + Haralick ─────────────────────────────────────────────────── */
    double *features = (double *)malloc((size_t)n_patches * N_FEATURES * sizeof(double));
    float  *glcm     = (float  *)malloc(GLCM_ELEMS * sizeof(float));

    for (int p = 0; p < n_patches; p++) {
        const uint8_t *patch = patches + (size_t)p * PATCH_SIZE * PATCH_SIZE;
        for (int a = 0; a < N_ANGLES; a++) {
            memset(glcm, 0, GLCM_ELEMS * sizeof(float));
            compute_glcm_patch(patch, PATCH_SIZE, PATCH_SIZE, a, glcm);
            double props[N_PROPS];
            haralick_props(glcm, props);
            for (int k = 0; k < N_PROPS; k++)
                features[(size_t)p * N_FEATURES + a * N_PROPS + k] = props[k];
        }
    }

    free(glcm);
    free(patches);

    double t2 = now_sec();

    /* ── Normalização + KMeans + Silhouette ─────────────────────────────── */
    standard_scale(features, n_patches, N_FEATURES);

    int *labels = (int *)malloc(n_patches * sizeof(int));
    double sil = kmeans_and_silhouette(features, n_patches, N_FEATURES,
                                        N_CLUSTERS, KMEANS_ITER, KMEANS_INIT,
                                        labels);
    free(labels);
    free(features);

    double t3 = now_sec();

    res.total_time_s      = t3 - t0;
    res.glcm_time_s       = t2 - t1;
    res.clustering_time_s = t3 - t2;
    res.silhouette_score  = sil;
    return res;
}

/* ── Main (uso standalone) ──────────────────────────────────────────────── */
#ifdef STANDALONE_MAIN
int main(int argc, char *argv[]) {
    const char *path = (argc > 1) ? argv[1] : "./data/image_512x512.bin";
    printf("CPU pipeline — %s\n", path);
    RunResult r = run_cpu_pipeline(path);
    printf("  total_time_s     : %.4f\n", r.total_time_s);
    printf("  glcm_time_s      : %.4f\n", r.glcm_time_s);
    printf("  clustering_time_s: %.4f\n", r.clustering_time_s);
    printf("  silhouette_score : %.4f\n", r.silhouette_score);
    printf("  n_patches        : %d\n",   r.n_patches);
    return 0;
}
#endif /* STANDALONE_MAIN */
